### ISSUE TITLE
Issue with `NotRequired` when instantiating a `TypedDict`

### ISSUE THREAD

**Issue by serramatutu** [microsoft/pyright#10340]
**Describe the bug**
Cannot assign an instance of a superclass `TypedDict` to a child class when it should be possible, i.e in the case the fields of the subclass are `NotRequired`.

In the following example, I'd expect `B(a)` to be perfectly valid since `{"a": 1}` is a valid value for `B` (`b` is `NotRequired`). However, I get the following error:

```
Argument of type "A" cannot be assigned to parameter "__map" of type "B" in function "__init__"
"b" is missing from "a"
```

**Code or Screenshots**
```python
from typing import NotRequired, TypedDict


class A(TypedDict):
    a: int

class B(A):
    b: NotRequired[int]

a = A({"a": 1})
b_ok = B({"a": 1})  # this is OK
b_err = B(a)        # this raises an error
```

**VS Code extension or command-line**
Using pyright on neovim. Version 1.1.399

**Comment by erictraut** [comment]
This behavior is correct and mandated by the typing spec, so I don't consider this a bug.

TypedDicts are structural types, and they are not "closed", which means that a dict can conform to a TypedDict definition but have additional keys. A value that conforms to `A` in your example can have an extra `b` of any type, so it is potentially not assignable to `B`.

```python
x: B = a # Not assignable according to the typing spec
```

The constructor for a TypedDict is overloaded and allows for several construction approaches. One accepts a value of the same type (in your example, the `B` constructor accepts another `B`) but also allows individual keyword arguments to specify overrides for individual keys and values. Using this overload, this is legal: `B(b_ok, b=1)`, but this is not: `B(a)`.

Another overload of the constructor supports only keyword arguments for the individual keys and values. Using this overload, this is legal: `B(**a)`. This is permitted because pyright and other type checkers are more permissive when it comes to unpack operations for argument expressions. There was a recent discussion about this in the [Python typing forum](https://discuss.python.org/t/standardizing-argument-unpacking-behaviors-for-calls/80747).

In any case, pyright is working as intended here, so I'm going to close this issue.

**Comment by serramatutu** [comment]
Thank you for the clarification!
